tensorboard/loader.py

```diff
@@ -0,0 +1,413 @@
+        self._wake_up_consumers.wait()
+      return self._get_record()
+
+  def close(self):
+    """Closes event log reader if open.
+
+    If the i/o thread is running, this method blocks until it has been
+    shut down.
+
+    Further reads are not permitted after this method is called.
+
+    Raises:
+      Exception: To propagate any exceptions that may have been thrown
+          by the close operation in the other thread. If an exception
+          is thrown, then all subsequent calls to this method will
+          rethrow that same exception.
+    """
+    with self._lock:
+      if not self._is_closed:
+        self._is_closed = True
+        if not self._thread.is_alive():
+          self._reader = None
+          return
+        self._wake_up_producer.notify()
+      while self._reader is not None:
+        self._wake_up_consumers.wait()
+      if self._close_exception is not None:
+        raise self._close_exception
+
+  def _get_record(self):
+    if self._read_exception is not None:
+      raise self._read_exception
+    if not self._records:
+      return None
+    record = self._records.popleft()
+    self._buffered -= len(record.record)
+    if self._should_wakeup():
+      self._wake_up_producer.notify()
+    return record
+
+  def _should_wakeup(self):
+    return (self._is_closed or
+            self._should_rebuffer() or
+            self._should_stat())
+
+  def _should_rebuffer(self):
+    return (not self._has_reached_end and
+            (float(self._buffered) <
+             self._read_ahead / BufferedRecordReader.READ_AHEAD_AGGRESSION))
+
+  def _should_stat(self):
+    return (self._offset > self._size or
+            self._last_stat <= self._clock() - self._stat_interval)
+
+  def _stat(self):
+    now = self._clock()
+    self._size = self._reader.get_size()
+    self._last_stat = now
+
+  def _run(self):
+    while True:
+      with self._lock:
+        while not self._should_wakeup():
+          self._wake_up_producer.wait()
+        if self._is_closed:
+          try:
+            self._reader.close()
+            tf.logging.debug('Closed')
+          except Exception as e:  # pylint: disable=broad-except
+            self._close_exception = e
+            tf.logging.debug('Close failed: %s', e)
+          self._reader = None
+          self._wake_up_consumers.notify_all()
+          return
+        # Calculate a good amount of data to read outside the lock.
+        # The less we have buffered, the less re-buffering we'll do.
+        # We want to minimize wait time in the other thread. See the
+        # following contour plot: https://goo.gl/HTBcCU
+        x = float(self._buffered)
+        y = BufferedRecordReader.READ_AHEAD_AGGRESSION
+        c = float(self._read_ahead)
+        want = int(min(c - x, y/c * x**y + 1))
+      # Perform re-buffering outside lock.
+      self._rebuffer(want)
+
+  def _rebuffer(self, want):
+    tf.logging.debug('Waking up to read %s bytes', util.add_commas(want))
```

Comment:
DEBUG isn't logged by default. It might be logged internally at Google. But even with DEBUG on, the math makes this log emitted surprisingly infrequently. It's great.